add(scan): Implement `RegisterKeys` service request to register new keys

upbqdn opened the pull request:
## Motivation

Close #8203.

### PR Author Checklist

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [x] Have you added or updated tests?
  - [x] Is the documentation up to date?

## Solution

- Preparation
  - [Refactor obtaining of activation heights](https://github.com/ZcashFoundation/zebra/commit/c7a7e6190739d7188f43d0c6b64fb27760c181f5)
  - [Mock viewing keys for tests](https://github.com/ZcashFoundation/zebra/commit/8ed93dda7fad09942b6cf42fd951d1f47e0a4b29)
- [Impl the RegisterKeys service request](https://github.com/ZcashFoundation/zebra/commit/c4a7246959b20a5ad1c7bc4675eb58cbe70142e0)
- [Refactor tests](https://github.com/ZcashFoundation/zebra/commit/9eb1cab8213f992cc3cc59987ca743fd9a0f5eb7)

## Review

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

upbqdn committed:
Refactor obtaining of activation heights

upbqdn committed:
Impl the `RegisterKeys` service request

upbqdn committed:
Mock viewing keys for tests

upbqdn committed:
Refactor tests

upbqdn committed:
Merge branch 'main' of github.com:ZcashFoundation/zebra

upbqdn commented:
`rustfmt` reported formatting errors in files this PR doesn't touch. Should be fixed in #8252.

arya2 commented on zebra-chain/src/parameters/network.rs:
```diff
@@ -129,6 +129,19 @@ impl Network {
         *self != Network::Mainnet
     }
+
+    /// Returns the minimum Sapling birthday height.
```
Optional/Nitpick:

```suggestion
    /// Returns the Sapling activation height for this network.
```

arya2 commented on zebra-scan/src/service/scan_task/tests/vectors.rs:
```diff
@@ -4,25 +4,23 @@ use std::collections::HashMap;
-            .map(|i| (i.to_string(), (vec![], vec![], Height::MIN)))
-            .collect(),
-    )?;
+    let sapling_keys = mock_sapling_viewing_keys(num_keys as u8);
```
This is much cleaner, thank you.

Optional/Nitpick:

```suggestion
    let sapling_keys = mock_sapling_viewing_keys(num_keys.try_into().expect("should fit in u8"));
```

upbqdn replied:
I thought `as` was a suitable shortcut here since it's not a production code, and `num_keys` is manually set one line above to a low value.

arya2 commented on zebra-scan/src/tests/vectors.rs:
```diff
@@ -146,18 +147,16 @@ async fn scanning_zecpages_from_populated_zebra_state() -> Result<()> {
 
     // Create a database
-    let mut s = new_test_storage(Network::Mainnet);
+    let mut s = new_test_storage(network);
```
Nitpick: This wasn't in this PR, and it's a test, but we should try to avoid single-letter variable names outside short closures.

upbqdn replied:
Addressed in [Avoid using a single-letter variable](https://github.com/ZcashFoundation/zebra/pull/8251/commits/5118f76a9199c2158f20fe9e91c915e538fcc88e).

arya2 commented on zebrad/tests/common/shielded_scan/scans_for_new_key.rs:
```diff
@@ -105,7 +105,12 @@ pub(crate) async fn run() -> Result<()> {
+    scan_task.register_keys(
+        parsed_keys
+            .into_iter()
+            .map(|key| (key.0, Some(key.1 .2 .0)))
```
Optional/Nitpick/Subjective: this may be slightly more clear if de-structured.

```suggestion
            .map(|(key, (_, _, Height(h)))| (key, Some(h)))
```

upbqdn replied:
I disagree. It's way more clear.

arya2 commented on zebrad/tests/common/shielded_scan/scans_for_new_key.rs:
No change needed: It might make refactors easier if we use service calls in acceptance tests instead of using ScanTask`.

arya2 reviewed (dismissed):
This looks excellent! Thank you!

All of my suggestions are optional, this PR looks great as-is.

upbqdn committed:
Merge branch 'main' into impl-key-registration-service

upbqdn committed:
Apply suggestions from code review

Co-authored-by: Arya <aryasolhi@gmail.com>

upbqdn committed:
Remove a redundant comment

I don't think we need to assume the genesis block doesn't contain
shielded data as the comment says.

upbqdn committed:
Avoid using a single-letter variable

upbqdn committed:
Refactor mocking Sapling scanning keys

upbqdn committed:
Merge branch 'main' into impl-key-registration-service

upbqdn commented:
I added two more changes that merit review:

- [Remove a redundant comment](https://github.com/ZcashFoundation/zebra/pull/8251/commits/fda282b5edcb3a3e5dfc589e217ca54a3ea048aa), and
- [Refactor mocking Sapling scanning keys](https://github.com/ZcashFoundation/zebra/pull/8251/commits/e634f192ded2efb3ee51076d1400300723385bc0).
